cpp/src/parquet/variant.h

```diff
@@ -0,0 +1,248 @@
+  std::string_view getMetadataKey(int32_t variant_id) const;
+  /// \brief Get the metadata id for a given key.
+  /// From the discussion in ML:
+  /// https://lists.apache.org/thread/b68tjmrjmy64mbv9dknpmqs28vnzjj96 if
+  /// !sortedStrings(), the metadata key is not guaranteed to be unique, so we use a
+  /// vector to store all the metadata ids.
+  ::arrow::internal::SmallVector<int32_t, 1> getMetadataId(std::string_view key) const;
+
+  bool sortedStrings() const;
+  uint8_t offsetSize() const;
+  uint32_t dictionarySize() const;
+
+ private:
+  static constexpr uint8_t VERSION_MASK = 0xF;
+  static constexpr uint8_t SORTED_STRING_MASK = 0b10000;
+
+ private:
+  std::string_view metadata_;
+};
+
+template <typename DecimalType>
+struct DecimalValue {
+  uint8_t scale;
+  DecimalType value;
+};
+
+struct VariantValue {
+  VariantMetadata metadata;
+  std::string_view value;
+
+  VariantBasicType getBasicType() const;
+  VariantType getType() const;
+  std::string typeDebugString() const;
+
+  /// \defgroup ValueAccessors
+  /// @{
+
+  // Note: Null doesn't need visitor.
+  bool getBool() const;
+  int8_t getInt8() const;
+  int16_t getInt16() const;
+  int32_t getInt32() const;
+  int64_t getInt64() const;
+  /// Include short_string optimization and primitive string type
+  std::string_view getString() const;
+  std::string_view getBinary() const;
+  float getFloat() const;
+  double getDouble() const;
+
+  DecimalValue<::arrow::Decimal32> getDecimal4() const;
+  DecimalValue<::arrow::Decimal64> getDecimal8() const;
+  DecimalValue<::arrow::Decimal128> getDecimal16() const;
+
+  int32_t getDate() const;
+  int64_t getTimeNTZ() const;
+  // timestamp with adjusted to UTC
+  int64_t getTimestamp() const;
+  int64_t getTimestampNTZ() const;
+  // 16 bytes UUID
+  std::array<uint8_t, 16> getUuid() const;
+
+  /// }@
+
+  struct ObjectInfo {
+    uint32_t num_elements;
+    uint32_t id_size;
+    uint32_t offset_size;
+    uint32_t id_start_offset;
+    uint32_t offset_start_offset;
+    uint32_t data_start_offset;
+
+    bool operator==(const ObjectInfo& info) const {
+      return num_elements == info.num_elements && id_size == info.id_size &&
+             offset_size == info.offset_size && id_start_offset == info.id_start_offset &&
+             offset_start_offset == info.offset_start_offset &&
+             data_start_offset == info.data_start_offset;
+    }
+
+    std::string toDebugString() const;
+  };
+  ObjectInfo getObjectInfo() const;
+  std::optional<VariantValue> getObjectValueByKey(std::string_view key) const;
+  std::optional<VariantValue> getObjectValueByKey(std::string_view key,
+                                                  const ObjectInfo& info) const;
+  std::optional<VariantValue> getObjectFieldByFieldId(uint32_t variant_id) const;
+  std::optional<VariantValue> getObjectFieldByFieldId(uint32_t variant_id,
+                                                      const ObjectInfo& info) const;
```

Comment:
Get with `ObjectInfo` is for performance, but they're so likely to produce bug. So I wonder can we mark it private or just leave it for performance..